src/box/box.cc

```diff
@@ -1805,6 +1805,234 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
+	if (vclock_count <= 1)
+		return 0;
+
+	struct sync_trigger_data data = {
+		.syncs = {0},
+		.collected_vclock_map = 0,
+		.waiter = fiber(),
+		.vclock = confirmed_vclock,
+		.deadline = deadline,
+		.count = vclock_count,
+		.is_timed_out = false,
+	};
+	struct trigger on_relay_thread_start;
+	trigger_create(&on_relay_thread_start, relay_get_sync_on_start, &data,
+		       NULL);
+	trigger_add(&replicaset.on_relay_thread_start, &on_relay_thread_start);
+	struct trigger on_ack;
+	trigger_create(&on_ack, check_sync_on_ack, &data, NULL);
+	trigger_add(&replicaset.on_ack, &on_ack);
+	replicaset_foreach(replica) {
+		if (relay_get_state(replica->relay) != RELAY_FOLLOW ||
+		    replica->anon) {
+			continue;
+		}
+		/* Might be already filled by on_relay_thread_start trigger. */
+		if (data.syncs[replica->id] != 0)
+			continue;
+		if (relay_next_sync(replica->relay, &data.syncs[replica->id],
+				    deadline) != 0) {
+			/* Timed out. */
+			trigger_clear(&on_ack);
+			trigger_clear(&on_relay_thread_start);
+			return -1;
+		}
+	}
+
+	while (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count &&
+	       !data.is_timed_out && !fiber_is_cancelled()) {
+		if (fiber_yield_deadline(deadline))
+			break;
+	}
+
+	trigger_clear(&on_ack);
+	trigger_clear(&on_relay_thread_start);
+	if (fiber_is_cancelled()) {
+		diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
+			 "fiber is cancelled");
+		return -1;
+	}
+	if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
+		diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
+			 "timed out waiting for remote node vclocks");
+		return -1;
+	}
+	return 0;
+}
+
+/** box_wait_vclock trigger data. */
+struct box_wait_vclock_data {
+	/** Whether the request is finished. */
+	bool is_ready;
+	/** The vclock to wait for. */
+	const struct vclock *vclock;
+	/** The fiber waiting for vclock. */
+	struct fiber *waiter;
+};
+
+static int
+box_wait_vclock_f(struct trigger *trigger, void *event)
+{
+	(void)event;
+	struct box_wait_vclock_data *data =
+		(struct box_wait_vclock_data *)trigger->data;
+	if (vclock_compare_ignore0(data->vclock, &replicaset.vclock) <= 0) {
+		data->is_ready = true;
+		fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/**
+ * Wait until this instance's vclock reaches @a vclock or @a deadline is
+ * reached.
+ */
+static int
+box_wait_vclock(const struct vclock *vclock, double deadline)
+{
+	if (vclock_compare_ignore0(vclock, &replicaset.vclock) <= 0)
+		return 0;
+	struct trigger on_wal_write;
+	struct box_wait_vclock_data data = {
+		.is_ready = false,
+		.vclock = vclock,
+		.waiter = fiber(),
+	};
+	trigger_create(&on_wal_write, box_wait_vclock_f, &data, NULL);
+	trigger_add(&wal_on_batch_write, &on_wal_write);
+	do {
+		if (fiber_yield_deadline(deadline))
+			break;
+	} while (!data.is_ready && !fiber_is_cancelled());
+	trigger_clear(&on_wal_write);
+	if (fiber_is_cancelled()) {
+		diag_set(ClientError, ER_VCLOCK_WAIT, vclock_to_string(vclock),
+			 "fiber is cancelled");
```

Comment:
Don't we want to be more verbose?

Anyway, I did that to conform to `box_wait_quorum`:
https://github.com/tarantool/tarantool/blob/60702571bec60c85d5e919e02dbf21bfa21de4e9/src/box/box.cc#L1794-L1797
https://github.com/tarantool/tarantool/blob/60702571bec60c85d5e919e02dbf21bfa21de4e9/src/box/box.cc#L1799-L1803

Should I change these places as well?

Reply:
> Don't we want to be more verbose?

I don't see a reason to be more verbose in case the fiber is cancelled. You can check file/line set in diag to figure out where exactly it was cancelled. I don't think that we should use different kinds of errors for this. BTW you use plain `ER_TIMEOUT` in case of timeout. Using plain `FiberIsCancelled` would only be consistent with that.

> https://github.com/tarantool/tarantool/blob/60702571bec60c85d5e919e02dbf21bfa21de4e9/src/box/box.cc#L1794-L1797

I'd change this place to use `FiberIsCancelled` as well.

Reply:
> I don't see a reason to be more verbose in case the fiber is cancelled. You can check file/line set in diag to figure out where exactly it was cancelled

I left a more detailed error in case of timeout. I think a user needs to know why the request timed out. Maybe the node was waiting for an unrealistic vclock or maybe it was waiting for too many responces.

<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index c79efd6e9..571803816 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1938,8 +1938,7 @@ box_collect_confirmed_vclock(struct vclock *confirmed_vclock, double deadline)
        trigger_clear(&on_ack);
        trigger_clear(&on_relay_thread_start);
        if (fiber_is_cancelled()) {
-               diag_set(ClientError, ER_QUORUM_WAIT, vclock_count,
-                        "fiber is cancelled");
+               diag_set(FiberIsCancelled);
                return -1;
        }
        if (bit_count_u32(data.collected_vclock_map) + 1 < vclock_count) {
@@ -1996,8 +1995,7 @@ box_wait_vclock(const struct vclock *vclock, double deadline)
        } while (!data.is_ready && !fiber_is_cancelled());
        trigger_clear(&on_wal_write);
        if (fiber_is_cancelled()) {
-               diag_set(ClientError, ER_VCLOCK_WAIT, vclock_to_string(vclock),
-                        "fiber is cancelled");
+               diag_set(FiberIsCancelled);
                return -1;
        }
        if (!data.is_ready) {

```

</details>